File: leetcode/python/695-max-area-of-island.py
# Author: RT
# Date: 2022-07-15T07:26:41.327Z
# URL: https://leetcode.com/problems/max-area-of-island/


class Solution:
    def maxAreaOfIsland(self, grid: list[list[int]]) -> int:
        ans = 0
        m, n = len(grid), len(grid[0])

        def dfs(i, j):
            if i < 0 or i >= m or j < 0 or j >= n:
                return 0
            if not grid[i][j]:
                return 0

            grid[i][j] = 0

            return 1 + dfs(i + 1, j) + dfs(i - 1, j) + dfs(i, j - 1) + dfs(i, j + 1)

        for i in range(m):
            for j in range(n):
                ans = max(ans, dfs(i, j))

        return ans

File: leetcode/rust/695-max-area-of-island.rs
// Author: RT
// Date: 2022-07-15T15:22:09.130Z
// URL: https://leetcode.com/problems/max-area-of-island/

impl Solution {
    pub fn max_area_of_island(mut grid: Vec<Vec<i32>>) -> i32 {
        let m = grid.len();
        let n = grid[0].len();

        let mut ans: i32 = 0;

        for i in 0..m {
            for j in 0..n {
                ans = std::cmp::max(ans, Self::dfs(&mut grid, i, j, m, n));
            }
        }

        ans
    }

    fn dfs(grid: &mut Vec<Vec<i32>>, i: usize, j: usize, m: usize, n: usize) -> i32 {
        if grid[i][j] == 0 {
            return 0;
        }

        grid[i][j] = 0;
        let mut area = 1;
        for dir in [0, 1, 0, -1, 0].windows(2) {
            let x = i as i32 + dir[0];
            let y = j as i32 + dir[1];
            if (0..m as i32).contains(&x) && (0..n as i32).contains(&y) {
                area += Self::dfs(grid, x as usize, y as usize, m, n);
            }
        }

        area
    }
}
